File: src/claude_memory/intelligence/entity_extraction.py
"""
Entity Extraction - Automatic entity identification and linking.

This module extracts entities from memory content using regex patterns
and optional NLP models. Supports file paths, functions, classes, errors,
technologies, concepts, and more.
"""

import re
import logging
from enum import Enum
from typing import Optional
from datetime import datetime
from pydantic import BaseModel, Field

logger = logging.getLogger(__name__)


class EntityType(Enum):
    """Types of entities that can be extracted from memory content."""

    FILE = "file"  # /path/to/file.py, file.txt
    FUNCTION = "function"  # function_name(), methodName()
    CLASS = "class"  # ClassName, ComponentName
    ERROR = "error"  # ErrorType, Exception, error codes
    TECHNOLOGY = "technology"  # Python, React, PostgreSQL
    CONCEPT = "concept"  # authentication, caching, CORS
    PERSON = "person"  # @username, developer names
    PROJECT = "project"  # project/repo names
    COMMAND = "command"  # CLI commands
    PACKAGE = "package"  # npm/pip package names
    URL = "url"  # HTTP(S) URLs
    VARIABLE = "variable"  # variable_name, CONSTANT_NAME


class Entity(BaseModel):
    """Represents an extracted entity."""

    text: str = Field(..., description="The extracted entity text")
    entity_type: EntityType = Field(..., description="Type of the entity")
    confidence: float = Field(default=1.0, ge=0.0, le=1.0, description="Extraction confidence")
    context: Optional[str] = Field(None, description="Surrounding context")
    start_pos: Optional[int] = Field(None, description="Start position in text")
    end_pos: Optional[int] = Field(None, description="End position in text")


class EntityExtractor:
    """Extracts entities from text using regex patterns."""

    # Regex patterns for different entity types
    PATTERNS = {
        EntityType.FILE: [
            # Absolute paths: /path/to/file.py
            r"(?:/[\w\-./]+)",
            # Relative paths with extension: src/file.py
            r"(?:[\w\-./]+\.[\w]+)",
            # Windows paths: C:\path\to\file.py
            r"(?:[A-Z]:\\[\w\-\\./]+)",
        ],
        EntityType.FUNCTION: [
            # function_name()
            r"\b([a-z_]\w*)\(\)",
            # methodName()
            r"\b([a-z]\w*[A-Z]\w*)\(\)",
        ],
        EntityType.CLASS: [
            # ClassName, Handler, Service, Manager, etc.
            r"\b([A-Z][\w]*(?:Class|Handler|Service|Manager|Controller|Provider|Factory|Builder|Strategy|Adapter|Facade|Proxy|Decorator|Observer|Singleton|Component|Module|Store|Action|Reducer|Hook|Context))\b",
            # Generic PascalCase
            r"\b([A-Z][a-z]+(?:[A-Z][a-z]+)+)\b",
        ],
        EntityType.ERROR: [
            # *Error, *Exception
            r"\b(\w*(?:Error|Exception))\b",
            # HTTP status codes
            r"\b([45]\d{2})\b",
            # Error codes like ERR_*, E_*
            r"\b(E(?:RR)?_[\w_]+)\b",
        ],
        EntityType.TECHNOLOGY: [
            # Programming languages
            r"\b(Python|JavaScript|TypeScript|Java|Kotlin|Swift|Go|Rust|C\+\+|C#|Ruby|PHP|Scala|Haskell|Elixir|Clojure|Erlang)\b",
            # Frameworks
            r"\b(React|Vue|Angular|Django|Flask|FastAPI|Express|Spring|Rails|Laravel|Symfony|Nest\.?js|Next\.?js|Nuxt\.?js|Svelte|Solid)\b",
            # Databases
            r"\b(PostgreSQL|MySQL|MongoDB|Redis|Neo4j|Memgraph|SQLite|DynamoDB|Cassandra|CouchDB|Elasticsearch|MariaDB|Oracle|MSSQL)\b",
            # Cloud/Infrastructure
            r"\b(AWS|Azure|GCP|Docker|Kubernetes|Terraform|Ansible|Jenkins|GitHub|GitLab|CircleCI|Travis)\b",
        ],
        EntityType.CONCEPT: [
            # Common programming concepts
            r"\b(authentication|authorization|caching|logging|testing|debugging|deployment|migration|refactoring|optimization|validation|serialization|deserialization|encryption|decryption|compression|decompression)\b",
            # Architecture patterns
            r"\b(MVC|MVVM|MVP|REST|GraphQL|gRPC|microservices|monolith|serverless|event-driven|CQRS|DDD|hexagonal|clean architecture)\b",
            # Security concepts
            r"\b(CORS|XSS|CSRF|SQL injection|JWT|OAuth|SAML|TLS|SSL|HTTPS|firewall|WAF)\b",
        ],
        EntityType.COMMAND: [
            # Commands in backticks or quotes
            r"`([^`]+)`",
            r'"([^"]+)"' + r'\s*(?:command|cmd|run|exec)',
        ],
        EntityType.PACKAGE: [
            # npm/pip packages
            r"\b((?:@[\w\-]+\/)?[\w\-]+)\b(?=\s*(?:package|library|module|dependency))",
            # Common package patterns
            r"\b(react-\w+|vue-\w+|@types/\w+|webpack-\w+|babel-\w+|eslint-\w+|pytest-\w+)\b",
        ],
        EntityType.URL: [
            # HTTP(S) URLs
            r"https?://[\w\-./]+(?:\?[\w\-=&]*)?",
        ],
        EntityType.VARIABLE: [
            # CONSTANT_NAME
            r"\b([A-Z][A-Z0-9_]{2,})\b",
            # snake_case
            r"\b([a-z_]\w*[a-z]\w*)\b(?=\s*[:=])",
        ],
    }

    def __init__(self, enable_nlp: bool = False):
        """
        Initialize the entity extractor.

        Args:
            enable_nlp: Enable NLP-based extraction (requires spaCy). Default: False
        """
        self.enable_nlp = enable_nlp
        self.nlp_model = None

        if enable_nlp:
            try:
                import spacy  # type: ignore

                self.nlp_model = spacy.load("en_core_web_sm")
                logger.info("NLP entity extraction enabled")
            except (ImportError, OSError):
                logger.warning(
                    "spaCy not available, falling back to regex-only extraction. "
                    "Install with: pip install spacy && python -m spacy download en_core_web_sm"
                )
                self.enable_nlp = False

    def extract(self, text: str, min_confidence: float = 0.5) -> list[Entity]:
        """
        Extract entities from text.

        Args:
            text: Text to extract entities from
            min_confidence: Minimum confidence threshold (0.0-1.0)

        Returns:
            List of extracted entities
        """
        entities: list[Entity] = []

        # Extract using regex patterns
        entities.extend(self._extract_with_regex(text))

        # Extract using NLP if enabled
        if self.enable_nlp and self.nlp_model:
            entities.extend(self._extract_with_nlp(text))

        # Deduplicate and filter by confidence
        entities = self._deduplicate(entities)
        entities = [e for e in entities if e.confidence >= min_confidence]

        return entities

    def _extract_with_regex(self, text: str) -> list[Entity]:
        """Extract entities using regex patterns."""
        entities: list[Entity] = []

        for entity_type, patterns in self.PATTERNS.items():
            for pattern in patterns:
                for match in re.finditer(pattern, text, re.IGNORECASE):
                    entity_text = match.group(1) if match.groups() else match.group(0)

                    # Skip very short or very long matches
                    if len(entity_text) < 2 or len(entity_text) > 100:
                        continue

                    # Calculate confidence based on pattern specificity
                    confidence = self._calculate_confidence(entity_type, entity_text, text)

                    # Extract context (50 chars before and after)
                    start = max(0, match.start() - 50)
                    end = min(len(text), match.end() + 50)
                    context = text[start:end]

                    entities.append(
                        Entity(
                            text=entity_text,
                            entity_type=entity_type,
                            confidence=confidence,
                            context=context,
                            start_pos=match.start(),
                            end_pos=match.end(),
                        )
                    )

        return entities

    def _extract_with_nlp(self, text: str) -> list[Entity]:
        """Extract entities using NLP (spaCy)."""
        entities: list[Entity] = []

        if not self.nlp_model:
            return entities

        doc = self.nlp_model(text)

        # Map spaCy entity types to our EntityType
        nlp_type_mapping = {
            "PERSON": EntityType.PERSON,
            "ORG": EntityType.PROJECT,  # Organizations often map to projects
            "PRODUCT": EntityType.TECHNOLOGY,
            "GPE": EntityType.CONCEPT,  # Geopolitical entities as concepts
        }

        for ent in doc.ents:
            if ent.label_ in nlp_type_mapping:
                entities.append(
                    Entity(
                        text=ent.text,
                        entity_type=nlp_type_mapping[ent.label_],
                        confidence=0.8,  # NLP confidence is generally high
                        context=ent.sent.text if ent.sent else None,
                        start_pos=ent.start_char,
                        end_pos=ent.end_char,
                    )
                )

        return entities

    def _calculate_confidence(self, entity_type: EntityType, text: str, full_text: str) -> float:
        """
        Calculate extraction confidence based on entity type and context.

        Args:
            entity_type: Type of entity
            text: Extracted entity text
            full_text: Full text being analyzed

        Returns:
            Confidence score (0.0-1.0)
        """
        confidence = 0.7  # Base confidence

        # Boost confidence for specific patterns
        if entity_type == EntityType.FILE:
            if text.endswith((".py", ".js", ".ts", ".jsx", ".tsx", ".md", ".txt", ".json", ".yaml", ".yml")):
                confidence = 0.95
            elif "/" in text or "\\" in text:
                confidence = 0.85

        elif entity_type == EntityType.FUNCTION:
            if "()" in text:
                confidence = 0.9

        elif entity_type == EntityType.CLASS:
            # Higher confidence for known suffixes
            if any(text.endswith(suffix) for suffix in ["Handler", "Service", "Manager", "Controller"]):
                confidence = 0.95
            else:
                confidence = 0.75

        elif entity_type == EntityType.ERROR:
            if text.endswith(("Error", "Exception")):
                confidence = 0.95
            elif re.match(r"[45]\d{2}", text):  # HTTP status codes
                confidence = 0.9

        elif entity_type == EntityType.TECHNOLOGY:
            # Known technologies have high confidence
            confidence = 0.95

        elif entity_type == EntityType.URL:
            confidence = 0.99

        elif entity_type == EntityType.COMMAND:
            # Commands in backticks are very reliable
            confidence = 0.9

        return min(confidence, 1.0)

    def _deduplicate(self, entities: list[Entity]) -> list[Entity]:
        """Remove duplicate entities, keeping highest confidence."""
        seen: dict[tuple[str, EntityType], Entity] = {}

        for entity in entities:
            key = (entity.text.lower(), entity.entity_type)

            if key not in seen or entity.confidence > seen[key].confidence:
                seen[key] = entity

        return list(seen.values())


# Singleton instance for convenience
_default_extractor = EntityExtractor()


def extract_entities(text: str, min_confidence: float = 0.5) -> list[Entity]:
    """
    Extract entities from text using the default extractor.

    Args:
        text: Text to extract entities from
        min_confidence: Minimum confidence threshold (0.0-1.0)

    Returns:
        List of extracted entities

    Example:
        >>> entities = extract_entities("Fixed authentication bug in src/auth.py")
        >>> for entity in entities:
        ...     print(f"{entity.entity_type.value}: {entity.text}")
        file: src/auth.py
        concept: authentication
    """
    return _default_extractor.extract(text, min_confidence)


async def link_entities(
    backend,
    memory_id: str,
    entities: list[Entity],
) -> list[str]:
    """
    Link extracted entities to a memory by creating entity nodes and MENTIONS relationships.

    Args:
        backend: Database backend instance
        memory_id: ID of the memory to link entities to
        entities: List of entities to link

    Returns:
        List of created entity IDs

    Example:
        >>> entities = extract_entities("Fixed React hooks issue")
        >>> entity_ids = await link_entities(backend, memory_id, entities)
    """
    entity_ids: list[str] = []

    for entity in entities:
        # Create or find entity node
        query = """
        MERGE (e:Entity {text: $text, type: $type})
        ON CREATE SET
            e.id = randomUUID(),
            e.created_at = datetime(),
            e.occurrence_count = 1
        ON MATCH SET
            e.occurrence_count = e.occurrence_count + 1,
            e.last_seen = datetime()
        WITH e
        MATCH (m:Memory {id: $memory_id})
        MERGE (m)-[r:MENTIONS]->(e)
        ON CREATE SET
            r.confidence = $confidence,
            r.created_at = datetime()
        RETURN e.id as entity_id
        """

        params = {
            "text": entity.text,
            "type": entity.entity_type.value,
            "memory_id": memory_id,
            "confidence": entity.confidence,
        }

        try:
            result = await backend.execute_query(query, params)
            if result:
                entity_ids.append(result[0]["entity_id"])
                logger.debug(
                    f"Linked entity '{entity.text}' ({entity.entity_type.value}) "
                    f"to memory {memory_id}"
                )
        except Exception as e:
            logger.error(f"Failed to link entity '{entity.text}': {e}")
            continue

    return entity_ids

File: ts/src/intelligence/entity-extraction.ts
/**
 * Entity Extraction - Automatic entity identification and linking.
 *
 * Port of the Python `memorygraph.intelligence.entity_extraction` module.
 * Extracts entities from memory content using regex patterns. The Python
 * version supports optional spaCy NLP extraction; spaCy is not available in
 * JS so NLP extraction is stubbed out and only regex-based extraction is
 * performed.
 */

import { z } from "zod";

import type { GraphBackend } from "../backends/index.js";

// ---------------------------------------------------------------------------
// Entity types
// ---------------------------------------------------------------------------

export const EntityType = {
  FILE: "file",
  FUNCTION: "function",
  CLASS: "class",
  ERROR: "error",
  TECHNOLOGY: "technology",
  CONCEPT: "concept",
  PERSON: "person",
  PROJECT: "project",
  COMMAND: "command",
  PACKAGE: "package",
  URL: "url",
  VARIABLE: "variable",
} as const;

export type EntityType = (typeof EntityType)[keyof typeof EntityType];

export const ALL_ENTITY_TYPES: string[] = Object.values(EntityType);

export function isEntityType(value: string): value is EntityType {
  return ALL_ENTITY_TYPES.includes(value);
}

// ---------------------------------------------------------------------------
// Entity schema / interface
// ---------------------------------------------------------------------------

export const EntitySchema = z.object({
  text: z.string().min(1),
  entity_type: z.enum(ALL_ENTITY_TYPES as [string, ...string[]]),
  confidence: z.number().min(0).max(1).default(1.0),
  context: z.string().nullish(),
  start_pos: z.number().int().nullish(),
  end_pos: z.number().int().nullish(),
});

export type Entity = z.infer<typeof EntitySchema>;

// ---------------------------------------------------------------------------
// Entity extractor
// ---------------------------------------------------------------------------

interface PatternSpec {
  /** Regex source. Flags are applied separately so we can capture groups. */
  source: string;
  flags: string;
  /**
   * Index of the capture group to use as the entity text. 0 means the entire
   * match (no capture group).
   */
  group: number;
}

const PATTERNS: Record<EntityType, PatternSpec[]> = {
  [EntityType.FILE]: [
    { source: "(?:/[\\w\\-./]+)", flags: "g", group: 0 },
    { source: "(?:[\\w\\-./]+\\.[\\w]+)", flags: "g", group: 0 },
    { source: "(?:[A-Z]:\\\\[\\w\\-\\\\./]+)", flags: "g", group: 0 },
  ],
  [EntityType.FUNCTION]: [
    { source: "\\b([a-z_]\\w*)\\(\\)", flags: "gi", group: 1 },
    { source: "\\b([a-z]\\w*[A-Z]\\w*)\\(\\)", flags: "g", group: 1 },
  ],
  [EntityType.CLASS]: [
    {
      source:
        "\\b([A-Z][\\w]*(?:Class|Handler|Service|Manager|Controller|Provider|Factory|Builder|Strategy|Adapter|Facade|Proxy|Decorator|Observer|Singleton|Component|Module|Store|Action|Reducer|Hook|Context))\\b",
      flags: "g",
      group: 1,
    },
    { source: "\\b([A-Z][a-z]+(?:[A-Z][a-z]+)+)\\b", flags: "g", group: 1 },
  ],
  [EntityType.ERROR]: [
    { source: "\\b(\\w*(?:Error|Exception))\\b", flags: "g", group: 1 },
    { source: "\\b([45]\\d{2})\\b", flags: "g", group: 1 },
    { source: "\\b(E(?:RR)?_[\\w_]+)\\b", flags: "g", group: 1 },
  ],
  [EntityType.TECHNOLOGY]: [
    {
      source:
        "\\b(Python|JavaScript|TypeScript|Java|Kotlin|Swift|Go|Rust|C\\+\\+|C#|Ruby|PHP|Scala|Haskell|Elixir|Clojure|Erlang)\\b",
      flags: "g",
      group: 1,
    },
    {
      source:
        "\\b(React|Vue|Angular|Django|Flask|FastAPI|Express|Spring|Rails|Laravel|Symfony|Nest\\.?js|Next\\.?js|Nuxt\\.?js|Svelte|Solid)\\b",
      flags: "g",
      group: 1,
    },
    {
      source:
        "\\b(PostgreSQL|MySQL|MongoDB|Redis|Neo4j|Memgraph|SQLite|DynamoDB|Cassandra|CouchDB|Elasticsearch|MariaDB|Oracle|MSSQL)\\b",
      flags: "g",
      group: 1,
    },
    {
      source:
        "\\b(AWS|Azure|GCP|Docker|Kubernetes|Terraform|Ansible|Jenkins|GitHub|GitLab|CircleCI|Travis)\\b",
      flags: "g",
      group: 1,
    },
  ],
  [EntityType.CONCEPT]: [
    {
      source:
        "\\b(authentication|authorization|caching|logging|testing|debugging|deployment|migration|refactoring|optimization|validation|serialization|deserialization|encryption|decryption|compression|decompression)\\b",
      flags: "g",
      group: 1,
    },
    {
      source:
        "\\b(MVC|MVVM|MVP|REST|GraphQL|gRPC|microservices|monolith|serverless|event-driven|CQRS|DDD|hexagonal|clean architecture)\\b",
      flags: "gi",
      group: 1,
    },
    {
      source: "\\b(CORS|XSS|CSRF|SQL injection|JWT|OAuth|SAML|TLS|SSL|HTTPS|firewall|WAF)\\b",
      flags: "g",
      group: 1,
    },
  ],
  [EntityType.COMMAND]: [
    { source: "`([^`]+)`", flags: "g", group: 1 },
    { source: '"([^"]+)"\\s*(?:command|cmd|run|exec)', flags: "gi", group: 1 },
  ],
  [EntityType.PACKAGE]: [
    {
      source: "\\b((?:@[\\w\\-]+\\/)?[\\w\\-]+)\\b(?=\\s*(?:package|library|module|dependency))",
      flags: "g",
      group: 1,
    },
    {
      source:
        "\\b(react-\\w+|vue-\\w+|@types/\\w+|webpack-\\w+|babel-\\w+|eslint-\\w+|pytest-\\w+)\\b",
      flags: "g",
      group: 1,
    },
  ],
  [EntityType.URL]: [
    {
      source: "https?://[\\w\\-./]+(?:\\?[\\w\\-=&]*)?",
      flags: "g",
      group: 0,
    },
  ],
  [EntityType.VARIABLE]: [
    { source: "\\b([A-Z][A-Z0-9_]{2,})\\b", flags: "g", group: 1 },
    { source: "\\b([a-z_]\\w*[a-z]\\w*)\\b(?=\\s*[:=])", flags: "g", group: 1 },
  ],
  [EntityType.PERSON]: [
    { source: "@([\\w\\-]+)", flags: "g", group: 1 },
  ],
  [EntityType.PROJECT]: [
    {
      source: "\\b([A-Z][a-z0-9]+(?:[A-Z][a-z0-9]+)*)\\b(?=\\s*(?:project|repo|repository))",
      flags: "g",
      group: 1,
    },
  ],
};

export class EntityExtractor {
  /** Enable NLP-based extraction. Stubbed - no JS spaCy equivalent. */
  enableNlp: boolean;

  constructor(enableNlp = false) {
    this.enableNlp = enableNlp;
  }

  /**
   * Extract entities from text.
   *
   * @param text - Text to extract entities from
   * @param minConfidence - Minimum confidence threshold (0.0-1.0)
   * @returns List of extracted entities
   */
  extract(text: string, minConfidence = 0.5): Entity[] {
    let entities: Entity[] = [];

    entities = entities.concat(this.extractWithRegex(text));

    // NLP extraction is stubbed out - no JS equivalent of spaCy available.
    if (this.enableNlp) {
      // Intentionally a no-op; logging kept for parity with Python version.
      console.warn(
        "NLP-based entity extraction is not available in the TypeScript port; using regex-only extraction."
      );
    }

    entities = this.deduplicate(entities);
    entities = entities.filter((e) => e.confidence >= minConfidence);

    return entities;
  }

  private extractWithRegex(text: string): Entity[] {
    const entities: Entity[] = [];

    for (const [entityTypeStr, patterns] of Object.entries(PATTERNS)) {
      const entityType = entityTypeStr as EntityType;
      for (const spec of patterns) {
        const re = new RegExp(spec.source, spec.flags);
        let match: RegExpExecArray | null;
        // Reset lastIndex in case the regex is stateful (it isn't here, but be safe).
        re.lastIndex = 0;
        while ((match = re.exec(text)) !== null) {
          const entityText = match[spec.group] ?? match[0];

          if (entityText.length < 2 || entityText.length > 100) {
            // Avoid infinite loops on zero-length matches.
            if (match.index === re.lastIndex) re.lastIndex++;
            continue;
          }

          const confidence = this.calculateConfidence(entityType, entityText);

          const start = Math.max(0, match.index - 50);
          const end = Math.min(text.length, match.index + entityText.length + 50);
          const context = text.slice(start, end);

          entities.push({
            text: entityText,
            entity_type: entityType,
            confidence,
            context,
            start_pos: match.index,
            end_pos: match.index + entityText.length,
          });

          // Guard against zero-length matches causing infinite loops.
          if (match.index === re.lastIndex) re.lastIndex++;
        }
      }
    }

    return entities;
  }

  private calculateConfidence(entityType: EntityType, text: string): number {
    let confidence = 0.7;

    if (entityType === EntityType.FILE) {
      if (
        text.endsWith(".py") ||
        text.endsWith(".js") ||
        text.endsWith(".ts") ||
        text.endsWith(".jsx") ||
        text.endsWith(".tsx") ||
        text.endsWith(".md") ||
        text.endsWith(".txt") ||
        text.endsWith(".json") ||
        text.endsWith(".yaml") ||
        text.endsWith(".yml")
      ) {
        confidence = 0.95;
      } else if (text.includes("/") || text.includes("\\")) {
        confidence = 0.85;
      }
    } else if (entityType === EntityType.FUNCTION) {
      if (text.endsWith("()")) confidence = 0.9;
    } else if (entityType === EntityType.CLASS) {
      if (
        text.endsWith("Handler") ||
        text.endsWith("Service") ||
        text.endsWith("Manager") ||
        text.endsWith("Controller")
      ) {
        confidence = 0.95;
      } else {
        confidence = 0.75;
      }
    } else if (entityType === EntityType.ERROR) {
      if (text.endsWith("Error") || text.endsWith("Exception")) {
        confidence = 0.95;
      } else if (/^[45]\d{2}$/.test(text)) {
        confidence = 0.9;
      }
    } else if (entityType === EntityType.TECHNOLOGY) {
      confidence = 0.95;
    } else if (entityType === EntityType.URL) {
      confidence = 0.99;
    } else if (entityType === EntityType.COMMAND) {
      confidence = 0.9;
    }

    return Math.min(confidence, 1.0);
  }

  private deduplicate(entities: Entity[]): Entity[] {
    const seen = new Map<string, Entity>();

    for (const entity of entities) {
      const key = `${entity.text.toLowerCase()}:${entity.entity_type}`;
      const existing = seen.get(key);
      if (!existing || entity.confidence > existing.confidence) {
        seen.set(key, entity);
      }
    }

    return Array.from(seen.values());
  }
}

// ---------------------------------------------------------------------------
// Singleton convenience
// ---------------------------------------------------------------------------

const defaultExtractor = new EntityExtractor();

export function extractEntities(text: string, minConfidence = 0.5): Entity[] {
  return defaultExtractor.extract(text, minConfidence);
}

// ---------------------------------------------------------------------------
// Entity linking
// ---------------------------------------------------------------------------

/**
 * Link extracted entities to a memory by creating entity nodes and MENTIONS
 * relationships.
 *
 * @param backend - Graph database backend instance
 * @param memoryId - ID of the memory to link entities to
 * @param entities - List of entities to link
 * @returns List of created entity IDs
 */
export async function linkEntities(
  backend: GraphBackend,
  memoryId: string,
  entities: Entity[]
): Promise<string[]> {
  const entityIds: string[] = [];

  for (const entity of entities) {
    const query = `
      MERGE (e:Entity {text: $text, type: $type})
      ON CREATE SET
        e.id = randomUUID(),
        e.created_at = datetime(),
        e.occurrence_count = 1
      ON MATCH SET
        e.occurrence_count = e.occurrence_count + 1,
        e.last_seen = datetime()
      WITH e
      MATCH (m:Memory {id: $memory_id})
      MERGE (m)-[r:MENTIONS]->(e)
      ON CREATE SET
        r.confidence = $confidence,
        r.created_at = datetime()
      RETURN e.id as entity_id
    `;

    const params = {
      text: entity.text,
      type: entity.entity_type,
      memory_id: memoryId,
      confidence: entity.confidence,
    };

    try {
      const result = await backend.executeQuery(query, params, true);
      if (result && result.length > 0) {
        const entityId = result[0]["entity_id"];
        if (typeof entityId === "string") {
          entityIds.push(entityId);
        }
      }
    } catch (e) {
      const message = e instanceof Error ? e.message : String(e);
      console.error(`Failed to link entity '${entity.text}': ${message}`);
      continue;
    }
  }

  return entityIds;
}
